src/installer/managed/Microsoft.NET.HostModel/ResourceUpdater.cs

```diff
@@ -352,126 +182,219 @@ public ResourceUpdater AddResource(byte[] data, string lpType, IntPtr lpName)
                 }
 
-                Debug.Assert(hResult != 0);
-                throw new HResultException(hResult);
+                // fix header
+                if (_reader.PEHeaders.PEHeader.Magic == PEMagic.PE32Plus)
+                {
+                    ModifyI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64InitializedDataSize,
+                        size => size + delta);
+                    ModifyI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64SizeOfImage,
+                        size => size + virtualDelta);
+
+                    if (needsMoveTrailingSections)
+                    {
+                        // fix RVA in DataDirectory
+                        for (int i = 0; i < _reader.PEHeaders.PEHeader.NumberOfRvaAndSizes; i++)
+                            PatchRVA(peSignatureOffset + Offsets.PEHeader.PE64DataDirectories + i * 8);
+                    }
+
+                    // index of ResourceTable is 2 in DataDirectories
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64DataDirectories + 2 * 8,
+                        rsrcVirtualAddress);
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64DataDirectories + 2 * 8 + 4,
+                        rsrcSectionDataSize);
+                }
+                else
+                {
+                    ModifyI32(buffer, peSignatureOffset + Offsets.PEHeader.PE32InitializedDataSize,
+                        size => size + delta);
+                    ModifyI32(buffer, peSignatureOffset + Offsets.PEHeader.PE32SizeOfImage,
+                        size => size + virtualDelta);
+
+                    if (needsMoveTrailingSections)
+                    {
+                        // fix RVA in DataDirectory
+                        for (int i = 0; i < _reader.PEHeaders.PEHeader.NumberOfRvaAndSizes; i++)
+                            PatchRVA(peSignatureOffset + Offsets.PEHeader.PE32DataDirectories + i * 8);
+                    }
+
+                    // index of ResourceTable is 2 in DataDirectories
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE32DataDirectories + 2 * 8,
+                        rsrcVirtualAddress);
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE32DataDirectories + 2 * 8 + 4,
+                        rsrcSectionDataSize);
+                }
             }
+
+            buffer.WriteArray(rsrcPointerToRawData, rsrcSectionData, 0, rsrcSectionDataSize);
+
+            // clear rest
+            //Array.Fill is standard 2.1
+            for (int i = rsrcSectionDataSize; i < newSectionSize; i++)
+                buffer.Write(rsrcPointerToRawData + i, (byte)0);
+
+            _resourceData = null;
         }
 
-        private static void CaptureEnumResourcesErrorInfo(IntPtr errorInfoPtr)
+        private static int ReadI32(MemoryMappedViewAccessor buffer, int position)
```

Comment:
MemoryMappedViewAccessor has `ReadInt32` and overloads of `Write` for `int` and `short` that you should be able to use instead of having these helpers

Reply:
It looks `ReadInt32` and `Write` uses native endianness so I reused helper function.